data_lang/utf8.h

```diff
@@ -0,0 +1,178 @@
+#ifndef DATA_LANG_UTF8_H
+#define DATA_LANG_UTF8_H
+
+#include <stdio.h>
+
+#include <stdint.h> // uint32_t
+#include <stddef.h> // size_t
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ *              ---- Quick reference about the encoding ----
+ *
+ * First, all valid UTF-8 sequences follow of bit "patterns" (Table 3-6.) The
+ * first byte determines the length of the sequence and then the next 0-3 bytes
+ * are "continuation bytes."
+ *
+ * +----------------------------+----------+----------+----------+----------+
+ * | Scalar Value               | 1st Byte | 2nd Byte | 3rd Byte | 4th Byte |
+ * +----------------------------+----------+----------+----------+----------+
+ * | 00000000 0xxxxxxx          | 0xxxxxxx |          |          |          |
+ * | 00000yyy yyxxxxxx          | 110yyyyy | 10xxxxxx |          |          |
+ * | zzzzyyyy yyxxxxxx          | 1110zzzz | 10yyyyyy | 10xxxxxx |          |
+ * | 000uuuuu zzzzyyyy yyxxxxxx | 11110uuu | 10uuzzzz | 10yyyyyy | 10xxxxxx |
+ * +----------------------------+----------+----------+----------+----------+
+ *
+ *      Table 3-6 from Unicode Standard 15.0.0 Ch3. UTF-8 bit patterns
+ *
+ * There are 3 further restrictions which make some valid bit patterns
+ * *invalid*:
+ *  1. Overlongs: eg, <0x41> and <0xC1 0x81> both store U+41, but the second
+ *     sequence is longer and thus an error.
+ *  2. Surrogates: Any codepoint between U+D800 and U+DFFF (inclusive) is a
+ *     surrogate. It is an error to encode surrogates in UTF-8.
+ *  3. Too Large: Any decoded value over 0x10FFFF is not a Unicode codepoint,
+ *     and must be rejected as an error.
+ *
+ * See https://aolsen.ca/writings/everything-about-utf8 for more details about
+ * the encoding.
+ */
+
+typedef enum Utf8Error {
+  UTF8_OK = 0,
+
+  // Encodes a codepoint in more bytes than necessary
+  UTF8_ERR_OVERLONG,
+
+  // Encodes a codepoint in the surrogate range (0xD800 to 0xDFFF, inclusive)
+  UTF8_ERR_SURROGATE,
+
+  // Encodes a value greater than the max codepoint U+10FFFF
+  UTF8_ERR_TOO_LARGE,
+
+  // Encoding doesn't conform to the UTF-8 bit patterns
+  UTF8_ERR_BAD_ENCODING,
+
+  // It looks like there is another codepoint, but it has been truncated.
+  UTF8_ERR_TRUNCATED_BYTES,
+
+  // We are at the end of the string. (input_len = 0)
+  UTF8_ERR_END_OF_STREAM,
+} Utf8Error_t;
+
+typedef struct Utf8Result {
+  Utf8Error_t error;
+  uint32_t codepoint;
+  size_t bytes_read;
+} Utf8Result_t;
+
+static inline void _cont(const unsigned char *input, Utf8Result_t *result) {
+  if (result->error) {
+    return;
+  }
+
+  int byte = input[result->bytes_read];
+  if (byte == '\0') {
+    result->error = UTF8_ERR_TRUNCATED_BYTES;
+    return;
+  }
+  result->bytes_read += 1;
+
+  // Continuation bytes follow the bit pattern 10xx_xxxx. We need to a)
+  // validate the pattern and b) remove the leading '10'.
+  if ((byte & 0xC0) == 0x80) {
+    result->codepoint <<= 6;
+    result->codepoint |= byte & 0x3F;
```

Comment:
( Too bad we can't use binary literals in C!  it's a lot easier to see that way )

Reply:
Just came out in C23^[ref](https://en.wikipedia.org/wiki/C23_(C_standard_revision)#Constants)! Should be able to start using them in a few decades :P